### ISSUE TITLE
`Install-Font`: File handle held on font by `GlyphTypeface` for process lifetime

### ISSUE THREAD

**Issue by akbyrd** [ralish/PSWinGlue#9]
I'm testing a script in Windows Sandbox and I've noticed that Install-Font can end up holding a file lock until the PowerShell instance exits. I've narrowed down the cause to this line
https://github.com/ralish/PSWinGlue/blob/ecdf8c48a579272c424e8ff71abb69b79bb9cbd8/Scripts/Install-Font.ps1#L205

I've not yet been able to find a workaround for this issue. The Shell method does not work properly in Sandbox, but Manual does work aside from the file lock.

Here's some janky example code I'm using
```pwsh
$tempDir = "fonts"
New-Item $tempDir -ItemType Directory -Force | Out-Null
Push-Location $tempDir
$progressPreference = "silentlyContinue"

$installFontUrl = "https://raw.githubusercontent.com/ralish/PSWinGlue/master/Scripts/Install-Font.ps1"
Invoke-WebRequest $installFontUrl -OutFile "Install-Font.ps1"

$fontUrl = "https://api.github.com/repos/ryanoasis/nerd-fonts/contents/patched-fonts/Agave"
$headers = @{ Accept = "application/vnd.github.raw" }
$result = iwr -UseBasicParsing -Headers $headers -Uri $fontUrl
$fontFiles = ConvertFrom-JSON $result.Content
foreach ($fontFile in $fontFiles)
{
	if ($fontFile.name.EndsWith(".ttf"))
	{
		Invoke-WebRequest -Uri $fontFile.download_url -OutFile $fontFile.name
	}
}

.\Install-Font.ps1 -Verbose . -Scope User -Method Manual

foreach ($fontFile in $fontFiles)
{
	if ($fontFile.name.EndsWith(".ttf"))
	{
		Remove-Item -Force $fontFile.name
	}
}
Remove-Item "Install-Font.ps1"

$progressPreference = "Continue"
Pop-Location
Remove-Item $tempDir
```

**Comment by ralish** [comment]
@akbyrd Hey Adam, thanks for the detailed report. Can you please post the output you receive as well, most importantly with the error/exception you receive when things go wrong?

**Comment by akbyrd** [comment]
Sure thing
```
VERBOSE: Installing font manually: AgaveNerdFont-Bold.ttf
Remove-Item : Cannot remove item C:\Users\WDAGUtilityAccount\fonts\AgaveNerdFont-Bold.ttf: Access to the path 'C:\Users\WDAGUtilityAccount\fonts\AgaveNerdFont-Bold.ttf' is denied.
At line:20 char:5
+     Remove-Item $fontFile.name
+     ~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : PermissionDenied: (C:\Users\WDAGUt...rdFont-Bold.ttf:FileInfo) [Remove-Item], UnauthorizedAccessException
    + FullyQualifiedErrorId : RemoveFileSystemItemUnAuthorizedAccess,Microsoft.PowerShell.Commands.RemoveItemCommand

Confirm
The item at C:\Users\WDAGUtilityAccount\fonts has children and the Recurse parameter was not specified. If you continue, all children will be removed with the item. Are you sure you want to continue?
[Y] Yes  [A] Yes to All  [N] No  [L] No to All  [S] Suspend  [?] Help (default is "Y"): y
Remove-Item : Cannot remove item C:\Users\WDAGUtilityAccount\fonts\AgaveNerdFont-Bold.ttf: Access to the path 'AgaveNerdFont-Bold.ttf' is denied.
At line:27 char:1
+ Remove-Item $tempDir
+ ~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : PermissionDenied: (AgaveNerdFont-Bold.ttf:FileInfo) [Remove-Item], UnauthorizedAccessException
    + FullyQualifiedErrorId : RemoveFileSystemItemUnAuthorizedAccess,Microsoft.PowerShell.Commands.RemoveItemCommand
Remove-Item : Directory C:\Users\WDAGUtilityAccount\fonts cannot be removed because it is not empty.
At line:27 char:1
+ Remove-Item $tempDir
+ ~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : WriteError: (C:\Users\WDAGUtilityAccount\fonts:DirectoryInfo) [Remove-Item], IOException
    + FullyQualifiedErrorId : DirectoryNotEmpty,Microsoft.PowerShell.Commands.RemoveItemCommand
```
![image](https://github.com/ralish/PSWinGlue/assets/484784/1ba4845c-257b-43bb-9351-4b6954a1d96a)

**Comment by ralish** [comment]
@akbyrd I tried reproducing this using your script and just experimenting on my own but haven't succeeded despite several attempts. A few thoughts:

1. What version of Windows are you using? I tested on Windows 11 22H2 x64 while you appear to be on some version of Windows 10. Feels unlikely this is related, but definitely not impossible.
2. Try disabling any antivirus temporarily. Again, feels unlikely, but AV is known for causing this class of issue. I note Windows Defender also is apparently "aware" of Windows Sandbox.

EDIT: Also curious is the line you identified doesn't appear to create any managed resources, of which opening a handle to a file would include. .NET types that do that implement the `IDisposable` interface, which the `GlyphTypeface` type does not. That again makes me suspicious there's something "external" that's influencing the normal process flow.

**Comment by akbyrd** [comment]
Sandbox reports Windows 10 Enterprise x64 2004 19041.3570. It doesn't appear to have antivirus enabled.

The host machine is running Windows 10 x64 Pro 22H2 19045.3570. I disabled "Real-time protection" in Windows Security and I'm able to reproduce the issue.

**Comment by ralish** [comment]
Interesting. Could you please perform a capture with [Process Monitor](https://learn.microsoft.com/en-us/sysinternals/downloads/procmon) and share it with me.

The general sequence of steps should be:

1. Open a new Windows Sandbox instance.
2. Download/copy Process Monitor into the instance and start it, accept the default capture filters, and confirm the capture has started (you should see lots of events).
3. Start PowerShell and reproduce the issue using the script you provided earlier. Once the issue has been reproduced, please wait a good 30 seconds and then attempt to remove the file again in PowerShell (I'm not expecting it to work, it just may prove helpful when analysing the capture).
4. Stop the capture and save it somewhere, then please Zip it and share with me a link to it using your service of choice (OneDrive, Dropbox, etc ...).

Cheers.

**Comment by akbyrd** [comment]
Here's the [trace](https://drive.google.com/file/d/1i7j4h5wlv7cFD4nzynEZ3Tn6MQlvlu2E/)

After I waited 30 seconds and tried to delete the file a second time (unsuccessfully) I closed PowerShell then deleted the file (successfully) from Windows Explorer. The file lock appears to be released the moment PowerShell is closed.

I wonder if the `GlyphTypeface` object doesn't get cleaned up immediately. It might not release the file until GC runs and that doesn't end up happening. Or something is holding onto a reference to that object and preventing it from being cleaned up. Just thinking out loud.

**Comment by akbyrd** [comment]
I also reproduced the issue with PowerShell 7.3.8 to rule out Windows PowerShell 5.1.19041.3570 as the cause.

**Comment by ralish** [comment]
Thanks @akbyrd, the logs were interesting but ultimately didn't reveal the underlying issue. However, I was able to reproduce the issue on a Windows 10 22H2 x64 VM and took a detailed look at the issue yesterday. Comparing the code paths taken for calls that open a file handle to the font against Windows 11 22H2 x64 shows nothing substantively different in the flow. In fact, all the function calls are identical. So it's not due to any difference in the .NET runtime, nor the system libraries it calls into.

It turns out that the same issue does affect Windows 11 22H2, in that the creation of the `GlyphTypeface` object results in a handle being created to the font which persists for the lifetime of the process, but it's effectively masked due to the usage of POSIX delete semantics. tl;dr: The latest Windows releases permit deleting a file even if it has open handles to it. Internally it's moved to a hidden directory on the filesystem, but this is all handled within NTFS, so is effectively invisible to the higher-level facilities `Install-Font` uses.

So where does that leave us? As far as I can tell the [DirectWrite](https://learn.microsoft.com/en-us/windows/win32/directwrite/direct-write-portal) library that `GlyphTypeface` uses internally opens a file handle and there's no way to close it from .NET. Forcing a garbage collection after disposing of all references to the object doesn't work, and the only exposed file object is via `GetFontStream()` which is unrelated. I think the best approach is to try creating the `GlyphTypeface` through an in-memory stream representing the font instead of the underlying file. It will almost certainly stay memory resident for the lifetime of the process (like it does now), but at least it won't lock the underlying file.

Doing this is non-trivial, but I'll take a look into it when I get the chance, hopefully this weekend.

**Comment by akbyrd** [comment]
That's some good sleuthing. Thanks for looking into it!

**Comment by ralish** [comment]
@akbyrd I've pushed an updated `Install-Font` which implements a new approach that passes the fonts to `GlyphTypeface` as memory streams instead of disk files. Could you test if it works for you? If it does, I'll publish an updated release to PSGallery.

**Comment by akbyrd** [comment]
It does indeed work now. Thanks :)

**Comment by ralish** [comment]
Published to the PowerShell Gallery as v0.6.8.
